### DISCUSSION TITLE
WebREPL sends 1 character at a time?

### DISCUSSION THREAD

**Discussion by aivarannamaa** [micropython/micropython#11686] [General]
It looks like WebREPL is sending output 1 character at a time -- i.e. makes a `send()` operation for each character.

Is it really so? If yes, then why? (If no, then I need to investigate my WebREPL client.)

**Comment by bulletmark** [comment]
Here is a response about this issue I wrote some time ago on the MP discord. Just copied and pasted here:

I have a personal CLI tool which I have used for a couple of years to develop MP code on various ESP32 devices around my house on wifi. The reason webrepl is so slow is because it is character based, i.e. each typed char is sent over the network socket and then echoed back by MP over the socket to the terminal. So I don't use the MP webrepl client. My implementation is line based using [websocket-client](https://github.com/websocket-client/websocket-client). It means I don't support RAW/paste mode etc, but I at least don't need that. Apart from being MUCH less laggy, another big advantage is that I can use GNU `readline` so have full command line history and editing (can use emacs or vi modes) which is also non-volatile (i.e. saved between invocations).

**Comment by mattytrentini** [comment]
A recent update to mpremote #11467 addresses this issue.

**Comment by cwalther** [comment]
> It looks like WebREPL is sending output 1 character at a time -- i.e. makes a `send()` operation for each character.
> Is it really so?

The question would be clearer if you could indicate which direction you are talking about – from MicroPython device (WebSocket server) to terminal (WebSocket client) or the other way around? Either way, I don’t know the answer offhand, but you should be able to figure it out by experimentation and reading the code of [either](https://github.com/micropython/micropython/blob/master/extmod/modwebrepl.c) [side](https://github.com/micropython/webrepl).

> A recent update to mpremote https://github.com/micropython/micropython/pull/11467 addresses this issue.

I’m confused, does mpremote now support WebREPL?

**Reply by mattytrentini** [reply]
No, I'm sorry, I complicated the issue! I was referring to a _similar_ issue with `mpremote`. Apologies!
